server/shadow/shadow_subsystem_builtin.c

```diff
@@ -22,70 +22,51 @@
 
 typedef struct
 {
-	const char* name;
+	const char* (*name)(void);
 	pfnShadowSubsystemEntry entry;
 } RDP_SHADOW_SUBSYSTEM;
 
-#ifdef WITH_SHADOW_X11
-extern int X11_ShadowSubsystemEntry(RDP_SHADOW_ENTRY_POINTS* pEntryPoints);
-#endif
-
-#ifdef WITH_SHADOW_MAC
-extern int Mac_ShadowSubsystemEntry(RDP_SHADOW_ENTRY_POINTS* pEntryPoints);
-#endif
-
-#ifdef WITH_SHADOW_WIN
-extern int Win_ShadowSubsystemEntry(RDP_SHADOW_ENTRY_POINTS* pEntryPoints);
-#endif
+extern int ShadowSubsystemEntry(RDP_SHADOW_ENTRY_POINTS* pEntryPoints);
+extern const char* ShadowSubsystemName(void);
 
 static RDP_SHADOW_SUBSYSTEM g_Subsystems[] = {
 
-#ifdef WITH_SHADOW_X11
-	{ "X11", X11_ShadowSubsystemEntry },
-#endif
-
-#ifdef WITH_SHADOW_MAC
-	{ "Mac", Mac_ShadowSubsystemEntry },
-#endif
-
-#ifdef WITH_SHADOW_WIN
-	{ "Win", Win_ShadowSubsystemEntry },
-#endif
-
-	{ "", NULL }
+	{ ShadowSubsystemName, ShadowSubsystemEntry }
 };
 
-static int g_SubsystemCount = (sizeof(g_Subsystems) / sizeof(g_Subsystems[0]));
+static size_t g_SubsystemCount = ARRAYSIZE(g_Subsystems);
 
 static pfnShadowSubsystemEntry shadow_subsystem_load_static_entry(const char* name)
 {
-	int index;
-
 	if (!name)
 	{
-		for (index = 0; index < g_SubsystemCount; index++)
+		for (size_t index = 0; index < g_SubsystemCount; index++)
```

Comment:
Doesn't this for loop become now superfluous (it does now always pick the first entry)?

Reply:
@pnowack well, it will return `NULL` if `g_Subsystem` is empty. but yes, not really required.